### PR TITLE
Implement marshaling of new integer types in C#

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR implements marshaling/unmarshaling for ushort, uint, ulong, varint, varlong, varuint and varulong, and makes the 2.0 encoding the default encoding.

### CODE REVIEW

File: csharp/src/Ice/IncomingRequestFrame.cs [hunk 1]
```diff
@@ -66,14 +66,15 @@ internal IncomingRequestFrame(Communicator communicator, OutgoingRequestFrame fr
...
+            InputStream.ReadEmptyEncapsulation(_communicator, Ice1Definitions.Encoding, Payload);
 
         /// <summary>Reads the request frame parameter list.</summary>
         /// <param name="reader">An InputStreamReader delegate used to read the request frame
         /// parameters.</param>
         /// <returns>The request parameters, when the frame parameter list contains multiple parameters
         /// they must be return as a tuple.</returns>
         public T ReadParamList<T>(InputStreamReader<T> reader) =>
-            InputStream.ReadEncapsulation(_communicator, Payload, reader);
+            InputStream.ReadEncapsulation(_communicator, Ice1Definitions.Encoding, Payload, reader);
```
Comment (bentoi) [inline csharp/src/Ice/IncomingRequestFrame.cs:78]: Ditto, I would change `ReadEncapsulation` here to `ReadEncapsulationPayload(_communicator, Encoding, Payload)`.

File: csharp/src/Ice/InputStream.cs [hunk 2]
```diff
@@ -217,6 +262,87 @@ public string ReadString()
...
+        }
+
+        /// <summary>Reads a ulong from the stream. This ulong is encoded using Ice's variable-length integer encoding.
+        /// </summary>
+        /// <returns>The ulong read from the stream.</returns>
+        public ulong ReadVarULong() => OldEncoding ? ReadULong() :
+            (_buffer[_pos] & 0x03) switch
+            {
+                0 => (uint)ReadByte() >> 2,
+                1 => (uint)ReadUShort() >> 2,
```
Comment (bentoi) [inline csharp/src/Ice/InputStream.cs:341]: It's not clear why we need the cast to uint here, perhaps add a comment?

File: csharp/src/Ice/InputStream.cs [hunk 3]
```diff
@@ -642,23 +768,25 @@ public object ReadSerializable()
...
         /// <param name="buffer">The byte buffer.</param>
         /// <param name="payloadReader">The reader used to read the payload of this encapsulation.</param>
-        internal static T ReadEncapsulation<T>(Communicator communicator,
-                                               ArraySegment<byte> buffer,
-                                               InputStreamReader<T> payloadReader)
+        internal static T ReadEncapsulation<T>(
+            Communicator communicator, Encoding encoding, ArraySegment<byte> buffer, InputStreamReader<T> payloadReader)
         {
-            var istr = new InputStream(communicator, buffer, startEncaps: true, 0);
+            var istr = new InputStream(communicator, encoding, buffer, startEncaps: true, 0);
```
Comment (bentoi) [inline csharp/src/Ice/InputStream.cs:789]: If we change this static method to ReadEncapsulationPayload we might be able to also get rid of the startEncaps parameter for the InputStream constructor. The check for whether or not the encoding is supported could directly be done here.

File: csharp/src/Ice/OutputStream.cs [hunk 4]
```diff
@@ -264,6 +324,90 @@ public void WriteString(string v)
             }
         }
 
+        /// <summary>Writes a uint to the stream.</summary>
+        /// <param name="v">The uint to write to the stream.</param>
+        public void WriteUInt(uint v) => WriteFixedSizeNumeric(v);
+
+        /// <summary>Writes a ulong to the stream.</summary>
+        /// <param name="v">The long to write to the stream.</param>
```
Comment (bentoi) [inline csharp/src/Ice/OutputStream.cs:332]:
```suggestion
        /// <param name="v">The ulong to write to the stream.</param>
```

File: csharp/src/Ice/OutputStream.cs [hunk 5]
```diff
@@ -264,6 +324,90 @@ public void WriteString(string v)
...
+        /// <param name="v">The ushort to write to the stream.</param>
+        public void WriteUShort(ushort v) => WriteFixedSizeNumeric(v);
+
+        /// <summary>Writes an int to stream, using Ice's variable-length encoding.</summary>
+        /// <param name="v">The int to write to the stream.</param>
+        public void WriteVarInt(int v) => WriteVarLong(v);
+
+        /// <summary>Writes a long to stream, using Ice's variable-length encoding, with the minimum number of bytes
+        /// required by the encoding.</summary>
+        /// <param name="v">The long to write to the stream. It must be in the range [2^61..2^61 - 1].</param>
```
Comment (bentoi) [inline csharp/src/Ice/OutputStream.cs:345]:
```suggestion
        /// <param name="v">The long to write to the stream. It must be in the range [-2^61..2^61 - 1].</param>
```

File: csharp/src/Ice/OutputStream.cs [hunk 6]
```diff
@@ -264,6 +324,90 @@ public void WriteString(string v)
...
+                };
+
+                v |= (uint)encodedLength;
+                Span<byte> data = stackalloc byte[sizeof(long)];
+                MemoryMarshal.Write(data, ref v);
+                WriteByteSpan(data.Slice(0, 1 << encodedLength));
+            }
+        }
+
+        /// <summary>Writes a uint to stream, using Ice's variable-length encoding.</summary>
```
Comment (bentoi) [inline csharp/src/Ice/OutputStream.cs:375]: "variable-length encoding" or "variable-length integer encoding", InputStream uses the later.

Comment (bernardnormier) [reply csharp/src/Ice/OutputStream.cs:375]: Fixed.

File: csharp/src/Ice/OutputStream.cs (hunk 5 above)
Comment (bernardnormier) [reply csharp/src/Ice/OutputStream.cs:345]: Fixed.

File: csharp/src/Ice/OutputStream.cs (hunk 4 above)
Comment (bernardnormier) [reply csharp/src/Ice/OutputStream.cs:332]: Fixed.

File: csharp/src/Ice/IncomingRequestFrame.cs [hunk 7]
```diff
@@ -66,14 +66,15 @@ internal IncomingRequestFrame(Communicator communicator, OutgoingRequestFrame fr
 
         /// <summary>Reads the empty parameter list, calling this methods ensure that the frame payload
         /// correspond to the empty parameter list.</summary>
-        public void ReadEmptyParamList() => InputStream.ReadEmptyEncapsulation(_communicator, Payload);
+        public void ReadEmptyParamList() =>
+            InputStream.ReadEmptyEncapsulation(_communicator, Ice1Definitions.Encoding, Payload);
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/IncomingRequestFrame.cs:70]: This is not directly to this PR, so I won't address it now.

Keep in mind that Payload points to the beginning of the encaps, so if we don't read the encaps header, we still need to skip it. For this to work properly, we would need to pass both the frame's header encoding (to know how many bytes to skip) and the payload encoding. And with the Ice 2.0's var-length size encoding, there won't be much difference between skipping the header and reading it (it's not that different with 1.1 either).

Reading the encaps header in IncomingResponseFrame/IncomingRequestFrame is primarily a sanity check during frame construction. For IncomingRequestFrame, it's also useful to figure out which encoding to use for the corresponding response. The payload Encoding of IncomingResponseFrame may not be needed beyond this initial sanity check.

File: csharp/src/Ice/InputStream.cs (hunk 2 above)
Comment (bernardnormier) [reply csharp/src/Ice/InputStream.cs:341]: Added comment.

**Comment by bernardnormier** [comment]
> Should we add tests to test out of range values?

The initial commit includes tests for out-of-range varlong and varulong in Twoways.cs (of the operations test).

The other types can't go out-of-range since the range for varint/varuint are the same as for int/uint.
